hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java

```diff
@@ -0,0 +1,368 @@
+              .getBucketTable().get(dbBucketKey);
+
+          if (bucketInfo == null) {
+            throw new OMException("Bucket " + snapInfo.getBucketName() +
+                " is not found", BUCKET_NOT_FOUND);
+          }
+
+          //TODO: [SNAPSHOT] Add lock to deletedTable and Active DB.
+          SnapshotInfo previousSnapshot = getPreviousSnapshot(snapInfo);
+          Table<String, OmKeyInfo> previousKeyTable = null;
+          OmSnapshot omPreviousSnapshot = null;
+
+          // Handle case when the deleted snapshot is the first snapshot.
+          // Move deleted keys to activeDB's deletedKeyTable
+          if (previousSnapshot != null) {
+            omPreviousSnapshot = (OmSnapshot) omSnapshotManager
+                .checkForSnapshot(previousSnapshot.getVolumeName(),
+                    previousSnapshot.getBucketName(),
+                    getSnapshotPrefix(previousSnapshot.getName()));
+
+            previousKeyTable = omPreviousSnapshot
+                .getMetadataManager().getKeyTable(bucketInfo.getBucketLayout());
+          }
+
+          // Move key to either next non deleted snapshot's snapshotDeletedTable
+          // or move to active object store deleted table
+
+          List<SnapshotMoveKeyInfos> toActiveDBList = new ArrayList<>();
+          List<SnapshotMoveKeyInfos> toNextDBList = new ArrayList<>();
+
+          try (TableIterator<String, ? extends Table.KeyValue<String,
+              RepeatedOmKeyInfo>> deletedIterator = snapshotDeletedTable
+              .iterator()) {
+
+            String snapshotBucketKey = dbBucketKey + OzoneConsts.OM_KEY_PREFIX;
+            iterator.seek(snapshotBucketKey);
+
+            while (deletedIterator.hasNext()) {
+              Table.KeyValue<String, RepeatedOmKeyInfo>
+                  deletedKeyValue = deletedIterator.next();
+              String deletedKey = deletedKeyValue.getKey();
+
+              // Exit if it is out of the bucket scope.
+              if (!deletedKey.contains(snapshotBucketKey)) {
+                break;
+              }
+
+              RepeatedOmKeyInfo repeatedOmKeyInfo = deletedKeyValue.getValue();
+
+              SnapshotMoveKeyInfos.Builder toActiveDb = SnapshotMoveKeyInfos
+                  .newBuilder()
+                  .setKey(deletedKey);
+              SnapshotMoveKeyInfos.Builder toNextDb = SnapshotMoveKeyInfos
+                  .newBuilder()
+                  .setKey(deletedKey);
+
+              for (OmKeyInfo keyInfo: repeatedOmKeyInfo.getOmKeyInfoList()) {
+                splitRepeatedOmKeyInfo(toActiveDb, toNextDb,
+                    keyInfo, previousKeyTable);
+              }
+
+              toActiveDBList.add(toActiveDb.build());
+              toNextDBList.add(toNextDb.build());
+
+            }
+            // Submit Move request to OM.
+            submitSnapshotMoveDeletedKeys(snapInfo, toActiveDBList,
+                toNextDBList);
+            snapshotLimit--;
```

Comment:
Do we need to change the snapshotInfo status to SNAPSHOT_RECLAIMED here?

Reply:
Thanks for pointing that out. Yes, we should be, but I'm putting it on another patch.